aptos-move/move-examples/mint_nft_v2/3-Adding-Admin/sources/create_nft_with_resource_and_admin_accounts.move

```diff
@@ -0,0 +1,201 @@
+        // ensure the signer attempting to change the admin is the current admin
+        assert!(current_admin_addr == mint_configuration.admin, error::permission_denied(ENOT_AUTHORIZED));
+        // ensure the new admin address is an account that's been initialized so we don't accidentally lock ourselves out
+        assert!(account::exists_at(new_admin_addr), error::not_found(ENOT_FOUND));
+        mint_configuration.admin = new_admin_addr;
+    }
+
+    public entry fun add_to_whitelist(
+        admin: &signer,
+        addresses: vector<address>,
+        resource_addr: address
+    ) acquires MintConfiguration {
+        let admin_addr = signer::address_of(admin);
+        let mint_configuration = borrow_global_mut<MintConfiguration>(resource_addr);
+        assert!(admin_addr == mint_configuration.admin, error::permission_denied(ENOT_AUTHORIZED));
+
+        vector::for_each(addresses, |user_addr| {
+            // note that this will abort in `table` if the address exists already- use `upsert` to ignore this
+            table::add(&mut mint_configuration.whitelist, user_addr, true);
+        });
+    }
+
+    public entry fun remove_from_whitelist(
+        admin: &signer,
+        addresses: vector<address>,
+        resource_addr: address
+    ) acquires MintConfiguration {
+        let admin_addr = signer::address_of(admin);
+        let mint_configuration = borrow_global_mut<MintConfiguration>(resource_addr);
+        assert!(admin_addr == mint_configuration.admin, error::permission_denied(ENOT_AUTHORIZED));
+
+        vector::for_each(addresses, |user_addr| {
+            // note that this will abort in `table` if the address is not found
+            table::remove(&mut mint_configuration.whitelist, user_addr);
+        });
+    }
+    
+    public entry fun set_minting_enabled(
+        admin: &signer,
+        minting_enabled: bool,
+        resource_addr: address,
+    ) acquires MintConfiguration {
+        let mint_configuration = borrow_global_mut<MintConfiguration>(resource_addr);
+        let admin_addr = signer::address_of(admin);
+        // abort if the signer is not the admin
+        assert!(admin_addr == mint_configuration.admin, error::permission_denied(ENOT_AUTHORIZED));
+        mint_configuration.minting_enabled = minting_enabled;
+    }
+
+    public entry fun set_expiration_timestamp(
+        admin: &signer,
+        expiration_timestamp: u64,
+        resource_addr: address,
+    ) acquires MintConfiguration {
+        let mint_configuration = borrow_global_mut<MintConfiguration>(resource_addr);
+        let admin_addr = signer::address_of(admin);
+        // abort if the signer is not the admin
+        assert!(admin_addr == mint_configuration.admin, error::permission_denied(ENOT_AUTHORIZED));
+        mint_configuration.expiration_timestamp = expiration_timestamp;
+    }
+
+    #[view]
+    public fun get_resource_address(collection_name: String): address {
```

Comment:
Shouldn't this just be @mint_nft_v2_part3?

Reply:
No because I create the resource account with a seed
 
```
        assert!(owner_addr == @mint_nft_v2_part2, error::permission_denied(ENOT_AUTHORIZED));

        let seed = *string::bytes(&collection_name);
        let (resource_signer, resource_signer_cap) = account::create_resource_account(owner, seed);
```

I could use an empty seed, but I always try to make sure the seed + signer (for creating the resource account) is a unique combination as a good habit, in this case creator + collection because I assert the creator is also the module deployer

But mainly this is just due to not using the `publish module to resource account` flow

Reply:
I still think we should publish the module to the resource account in this example. Users having to pass in the resource address reminds me of one similar pain point in Sui where users have to pass in the object address where the contract state is stored. This is a rather painful UX IMO as the addresses would need to be hardcoded in all off chain components (FE, indexing, explorer, etc.) for users to interact or find the data associated.

I think it's worthwhile in our system to eliminate this pain if possible so that common data of a contract is stored along with the module's code and is thus much easier to find and interact with

Reply:
Okay so I thought through this-originally with an admin model I thought that you could lose track of the original creator, in which case you'd still need to provide the original creator address + collection name.

However that's only the case if the module supports multiple collections from different "owners" otherwise you only need the collection name (and not even that if the module only supports a single collection), because if you restrict it to a single "owner" you can just store that information at the contract address